File: compilers/java/src/test/java/runar/compiler/passes/C18PrivateVarLenStateReadTest.java
package runar.compiler.passes;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotEquals;

import org.junit.jupiter.api.Test;

/**
 * Regression test for deep-review finding C18 (P1 funds-safety bug).
 *
 * <p>{@code StackLower.methodReadsVarLenState} used to walk only a method's OWN
 * bindings (plus {@code if}/{@code loop} bodies) looking for a direct
 * {@code load_prop} on a mutable variable-length (ByteString) state field. It did
 * NOT recurse into private helper methods invoked via {@code method_call} — unlike
 * its sibling walker {@code methodUsesCheckPreimage}, which already did.
 *
 * <p>Consequence: a PUBLIC method that reads a mutable ByteString state field ONLY
 * through a PRIVATE helper never set {@code usesCodePart}, so {@code _codePart} was
 * never pushed as an implicit parameter and the var-length deserialization was
 * skipped. The private helper's {@code load_prop} then fell back to the deploy-time
 * constant instead of the live on-chain state — a silent wrong-result bug.
 */
class C18PrivateVarLenStateReadTest {

    /** Control: the public method reads the mutable ByteString field directly. */
    private static final String DIRECT =
        "import { StatefulSmartContract, assert, len } from 'runar-lang';\n"
        + "import type { ByteString } from 'runar-lang';\n"
        + "class StateReadDirect extends StatefulSmartContract {\n"
        + "  tag: ByteString;\n"
        + "  constructor(tag: ByteString) { super(tag); this.tag = tag; }\n"
        + "  public check(expected: bigint): void { assert(len(this.tag) === expected); }\n"
        + "}\n";

    /** Bug case: the public method reads the same field ONLY via a private helper. */
    private static final String VIA_HELPER =
        "import { StatefulSmartContract, assert, len } from 'runar-lang';\n"
        + "import type { ByteString } from 'runar-lang';\n"
        + "class StateReadDirect extends StatefulSmartContract {\n"
        + "  tag: ByteString;\n"
        + "  constructor(tag: ByteString) { super(tag); this.tag = tag; }\n"
        + "  private tagLen(): bigint { return len(this.tag); }\n"
        + "  public check(expected: bigint): void { assert(this.tagLen() === expected); }\n"
        + "}\n";

    /** Two private helpers calling each other must not hang the cycle guard. */
    private static final String VIA_HELPER_CHAIN =
        "import { StatefulSmartContract, assert, len } from 'runar-lang';\n"
        + "import type { ByteString } from 'runar-lang';\n"
        + "class StateReadDirect extends StatefulSmartContract {\n"
        + "  tag: ByteString;\n"
        + "  constructor(tag: ByteString) { super(tag); this.tag = tag; }\n"
        + "  private inner(): bigint { return len(this.tag); }\n"
        + "  private outer(): bigint { return this.inner(); }\n"
        + "  public check(expected: bigint): void { assert(this.outer() === expected); }\n"
        + "}\n";

    /** No var-length read at all — must keep the original terminal codegen. */
    private static final String NO_VARLEN_READ =
        "import { StatefulSmartContract, assert } from 'runar-lang';\n"
        + "import type { ByteString } from 'runar-lang';\n"
        + "class StateReadDirect extends StatefulSmartContract {\n"
        + "  tag: ByteString;\n"
        + "  constructor(tag: ByteString) { super(tag); this.tag = tag; }\n"
        + "  private twice(x: bigint): bigint { return x + x; }\n"
        + "  public check(expected: bigint): void { assert(this.twice(expected) === expected + expected); }\n"
        + "}\n";

    @Test
    void readViaPrivateHelperMatchesDirectRead() throws Exception {
        String direct = PipelineTestSupport.hex(DIRECT, "StateReadDirect.runar.ts");
        String viaHelper = PipelineTestSupport.hex(VIA_HELPER, "StateReadDirect.runar.ts");
        assertEquals(direct, viaHelper,
            "reading a mutable ByteString field through a private helper must lower "
            + "identically to reading it directly (both need _codePart)");
    }

    @Test
    void readViaChainedPrivateHelpersMatchesDirectRead() throws Exception {
        String direct = PipelineTestSupport.hex(DIRECT, "StateReadDirect.runar.ts");
        String viaChain = PipelineTestSupport.hex(VIA_HELPER_CHAIN, "StateReadDirect.runar.ts");
        assertEquals(direct, viaChain,
            "transitive private-helper reads must also force _codePart");
    }

    @Test
    void methodWithoutVarLenReadKeepsTerminalCodegen() throws Exception {
        String direct = PipelineTestSupport.hex(DIRECT, "StateReadDirect.runar.ts");
        String noRead = PipelineTestSupport.hex(NO_VARLEN_READ, "StateReadDirect.runar.ts");
        assertNotEquals(direct, noRead,
            "a method that never reads the mutable ByteString field must NOT gain _codePart");
    }
}

File: packages/runar-compiler/src/__tests__/c18-private-varlen-state-read.test.ts
/**
 * Regression test for deep-review finding C18 (P1 funds-safety bug).
 *
 * `usesCodePart` / `methodReadsVarLenState` in `05-stack-lower.ts` only walks
 * a method's OWN bindings (plus `if`/`loop` bodies) looking for a direct
 * `load_prop` on a mutable variable-length (ByteString) state field. It does
 * NOT recurse into private helper methods invoked via `method_call` — unlike
 * its sibling walker `methodUsesCheckPreimage`, which already does this.
 *
 * Consequence: a PUBLIC method that reads a mutable ByteString state field
 * ONLY through a PRIVATE helper never sets `usesCodePart`, so `_codePart` is
 * never pushed as an implicit parameter and `lowerDeserializeState` takes the
 * "terminal method, skip variable-length deserialization" shortcut. The
 * private helper's `load_prop` then falls back to the deploy-time constant
 * (or constructor placeholder) instead of the live on-chain state — a silent
 * wrong-result / funds-safety bug for any call built against current state.
 */
import { describe, it, expect } from 'vitest';
import { parse } from '../passes/01-parse.js';
import { lowerToANF } from '../passes/04-anf-lower.js';
import { lowerToStack } from '../passes/05-stack-lower.js';
import { compile } from '../index.js';
import type { StackProgram } from '../ir/index.js';

function compileToStack(source: string): StackProgram {
  const parseResult = parse(source);
  if (!parseResult.contract) {
    throw new Error(`Parse failed: ${parseResult.errors.map(e => e.message).join(', ')}`);
  }
  const anf = lowerToANF(parseResult.contract);
  return lowerToStack(anf);
}

// Control: the public method reads the mutable ByteString field directly.
const DIRECT = `import { StatefulSmartContract, assert, len } from 'runar-lang';
import type { ByteString } from 'runar-lang';
class StateReadDirect extends StatefulSmartContract {
  tag: ByteString;
  constructor(tag: ByteString) { super(tag); this.tag = tag; }
  public check(expected: bigint): void { assert(len(this.tag) === expected); }
}`;

// Bug case: the public method reads the same field ONLY via a private helper.
const VIA_HELPER = `import { StatefulSmartContract, assert, len } from 'runar-lang';
import type { ByteString } from 'runar-lang';
class StateReadViaHelper extends StatefulSmartContract {
  tag: ByteString;
  constructor(tag: ByteString) { super(tag); this.tag = tag; }
  private tagLen(): bigint { return len(this.tag); }
  public check(expected: bigint): void { assert(this.tagLen() === expected); }
}`;

describe('C18 — private-helper read of mutable ByteString state must force _codePart', () => {
  it('control: direct read sets usesCodePart on the public method (Stack IR)', () => {
    const program = compileToStack(DIRECT);
    const check = program.methods.find(m => m.name === 'check');
    expect(check).toBeDefined();
    expect(check!.usesCodePart).toBe(true);
  });

  it('bug case: read via private helper must ALSO set usesCodePart on the public method (Stack IR)', () => {
    const program = compileToStack(VIA_HELPER);
    const check = program.methods.find(m => m.name === 'check');
    expect(check).toBeDefined();
    expect(check!.usesCodePart).toBe(true);
  });

  it('end-to-end: compiled artifact ABI reports usesCodePart identically for both variants', () => {
    const direct = compile(DIRECT, { fileName: 'StateReadDirect.runar.ts' });
    const viaHelper = compile(VIA_HELPER, { fileName: 'StateReadViaHelper.runar.ts' });

    expect(direct.success).toBe(true);
    expect(viaHelper.success).toBe(true);

    const directCheck = direct.artifact!.abi.methods.find(m => m.name === 'check');
    const viaHelperCheck = viaHelper.artifact!.abi.methods.find(m => m.name === 'check');
    expect(directCheck).toBeDefined();
    expect(viaHelperCheck).toBeDefined();

    expect(directCheck!.usesCodePart).toBe(true);
    expect(viaHelperCheck!.usesCodePart).toBe(true);
  });
});
